// inngest/components/execution/otel/aiExtractor.cjs

//#region src/components/execution/otel/aiExtractor.ts
/**
* A `convention` identifies a namespace of AI span attributes and orders which
* should win when multiple namespaces are present on the same span. Lower
* numbers override higher numbers.
*/
var Convention = /* @__PURE__ */ function(Convention$1) {
	Convention$1[Convention$1["Langfuse"] = 1] = "Langfuse";
	Convention$1[Convention$1["Semconv"] = 2] = "Semconv";
	Convention$1[Convention$1["OpenInference"] = 3] = "OpenInference";
	Convention$1[Convention$1["Vercel"] = 4] = "Vercel";
	return Convention$1;
}(Convention || {});
/**
* Prefix for the synthetic attributes produced by
* {@link expandLangfuseUsageDetails}, kept distinct from any real attribute
* namespace.
*/
const langfuseUsagePrefix = "__langfuse.usage_details.";
/**
* Parses the `langfuse.observation.usage_details` JSON blob (e.g.
* `{"input":17,"output":36,"total":53}`) and emits a synthetic scalar
* attribute for each count we track. Only `input` is emitted; the other counts
* (output, total, cached, reasoning, …) are intentionally dropped since this
* extractor tracks input tokens alone.
*/
const expandLangfuseUsageDetails = (value) => {
	if (typeof value !== "string" || value === "") return {};
	let counts;
	try {
		counts = JSON.parse(value);
	} catch {
		return {};
	}
	if (typeof counts !== "object" || counts === null) return {};
	const input = counts.input;
	if (typeof input !== "number") return {};
	return { [`${langfuseUsagePrefix}input`]: input };
};
/**
* Maps a source attribute key to the canonical {@link AIMetadata} field it
* populates and the convention it belongs to.
*/
const keyFieldMap = {
	"langfuse.observation.usage_details": {
		convention: Convention.Langfuse,
		expand: expandLangfuseUsageDetails
	},
	[`${langfuseUsagePrefix}input`]: {
		field: "inputTokens",
		convention: Convention.Langfuse
	},
	"gen_ai.request.model": {
		field: "model",
		convention: Convention.Semconv
	},
	"gen_ai.usage.input_tokens": {
		field: "inputTokens",
		convention: Convention.Semconv
	},
	"llm.model_name": {
		field: "model",
		convention: Convention.OpenInference
	},
	"llm.token_count.prompt": {
		field: "inputTokens",
		convention: Convention.OpenInference
	},
	"ai.model.id": {
		field: "model",
		convention: Convention.Vercel
	},
	"ai.usage.inputTokens": {
		field: "inputTokens",
		convention: Convention.Vercel
	},
	"ai.usage.tokens": {
		field: "inputTokens",
		convention: Convention.Vercel
	}
};
/**
* Extracts AI model metadata from a span's attributes.
*
* Attributes are matched across multiple instrumentation conventions
* (OpenTelemetry semconv, OpenInference, Vercel AI SDK). When more than one
* convention supplies the same field, the highest-precedence value wins.
*
* @param attributes - The span attributes, as exposed by
* `ReadableSpan.attributes`.
*/
const extractAIMetadataFromAttributes = (attributes) => {
	const candidates = {};
	const record = (mapping, value) => {
		if (!mapping.field) return;
		const existing = candidates[mapping.field];
		if (!existing || mapping.convention < existing.convention) candidates[mapping.field] = {
			value,
			convention: mapping.convention
		};
	};
	for (const [key, value] of Object.entries(attributes)) {
		if (value === void 0) continue;
		const mapping = keyFieldMap[key];
		if (!mapping) continue;
		if (mapping.expand) {
			for (const [childKey, childValue] of Object.entries(mapping.expand(value))) {
				const childMapping = keyFieldMap[childKey];
				if (childMapping) record(childMapping, childValue);
			}
			continue;
		}
		record(mapping, value);
	}
	const metadata = {};
	const model = candidates.model?.value;
	if (typeof model === "string" && model !== "") metadata.model = model;
	const inputTokens = candidates.inputTokens?.value;
	if (inputTokens !== void 0) {
		const n = Number(inputTokens);
		if (!Number.isNaN(n)) metadata.inputTokens = n;
	}
	return metadata;
};
/**
* Aggregates two {@link AIMetadata} values into one.
*
* Input token counts are summed, while `a`'s model takes precedence over `b`'s.
* Each field is only present in the result when at least one input supplies it.
*
* @param a - The primary metadata; its `model` wins when both are present.
* @param b - The secondary metadata.
*/
const aggregate = (a, b) => {
	const metadata = {};
	const model = a.model ?? b.model;
	if (model !== void 0) metadata.model = model;
	if (a.inputTokens !== void 0 || b.inputTokens !== void 0) metadata.inputTokens = (a.inputTokens ?? 0) + (b.inputTokens ?? 0);
	return metadata;
};
/**
* Maps an {@link AIMetadata} value onto the server's `inngest.ai` metadata
* schema (snake_case keys). Only the fields we extract are emitted; absent
* fields are omitted rather than zero-valued. Returns `undefined` when there
* is nothing to emit, so callers can skip stamping metadata entirely.
*/
const toInngestAIMetadataValues = (metadata) => {
	const values = {};
	if (metadata.model !== void 0) values.model = metadata.model;
	if (metadata.inputTokens !== void 0) values.input_tokens = metadata.inputTokens;
	if (Object.keys(values).length === 0) return;
	return values;
};

//#endregion
exports.aggregate = aggregate;
exports.extractAIMetadataFromAttributes = extractAIMetadataFromAttributes;
exports.toInngestAIMetadataValues = toInngestAIMetadataValues;
